apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+which bridged the functionality gap using a composition of dense matrix multiplications. Here we
+compare to both the dense baseline and these older implementations of the same operation.
+
+Again, we will look at the performance for [`float32`][#fig-4] first.  The plot compares the new
+triton-based implementation to the one available in older versions of PyTorch.  The triton backed
+kernels show significant improvements over the previous version. Speedup near 2x is observed at
+sparsity ratios near 50%. At high sparsity ratio it exceeds 6x. This is a leap forward in terms of
+raw sparse performance but we are primarily concerned with performance compared to the dense baseline
+as this will determine if sparsity can be used without a performance penalty.
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_float32_vs_sparse.png"
+    name="fig-4"
+    alt="Two plots displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup
+    over the baseline on the vertical.  Three lines are plotted on both upper and lower plots
+    indicating the block sizes used 16, 32, and 64."
+/>
+
+Comparison to the [dense baseline][#fig-5] is also very promising. Block sizes of 32 and 64 are
+faster than dense for all sparsity ratios, and block size 16 crossing the 1x boundary at 70% sparsity.
+This also represents a leap forward in the usability of sparse tensors. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_float32_vs_dense.png"
+    name="fig-5"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+Now we look at the performance for half precision data types. The new kernel provides between 20-50x
+speedup over the [older sparse implementation][#fig-6]. While a speedup of this size is always a notable
+result, we have known that the implementation supporting half precision data types would not be
+optimal. However, as cuSPARSE did not support these data types a working implementation was created
+by composing calls into the dense kernel for each sparse block. This approach suffers from a
+significant penalty due to launching many CUDA kernels. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_sparse.png"
+    name="fig-6"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  Three lines are plotted indicating block sizes
+    used for the sparse tensor of 16, 32, and 64."
+/>
+
+Comparing performance for half precision to the [dense baseline][#fig-7] we again see one clear win.
```

Comment:
"one clear win" or "a clear win"? I don't have the plots here, so I don't know which one is it.